File: server/internal/game/roll_coverage.go
package game

import (
	"fmt"
	"sort"
)

// ─── Roll coverage ──────────────────────────────────────────────────────────
//
// Weighted lists and loot tables share one rule: their ranges must TILE the die
// — cover 1..maxRoll with no gaps and no overlaps, so every possible roll lands
// on exactly one outcome.
//
// Gaps are an ERROR, not an implicit "nothing happens". A no-drop chance is a
// thing you say out loud (a table's `nothing` row), because a hole in the ranges
// is indistinguishable from a typo — and the one gap that ever shipped
// (raider_loot's 51-60, a deliberate 10% no-drop) read exactly like one.

// rollRange is one claimed slice of a die, labelled for error messages.
type rollRange struct {
	Min, Max int
	Label    string // what this range yields, e.g. an item id or "nothing"
}

// validateRollCoverage checks that ranges tile 1..maxRoll exactly. `what` and
// `id` only shape the error message (e.g. "list", "basic_weapons").
func validateRollCoverage(what, id string, maxRoll int, ranges []rollRange) error {
	if maxRoll < 1 {
		return fmt.Errorf("%s %q: maxRoll must be at least 1, got %d", what, id, maxRoll)
	}
	if len(ranges) == 0 {
		return fmt.Errorf("%s %q: needs at least 1 entry", what, id)
	}

	for _, r := range ranges {
		if r.Min < 1 || r.Max > maxRoll {
			return fmt.Errorf("%s %q: %s claims rolls %d-%d, outside the die 1-%d",
				what, id, r.Label, r.Min, r.Max, maxRoll)
		}
		if r.Min > r.Max {
			return fmt.Errorf("%s %q: %s claims rolls %d-%d, which is backwards",
				what, id, r.Label, r.Min, r.Max)
		}
	}

	// Sort a copy so the author's authored order is not a correctness constraint.
	sorted := append([]rollRange(nil), ranges...)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Min < sorted[j].Min })

	next := 1 // the lowest roll not yet claimed
	for _, r := range sorted {
		if r.Min > next {
			return fmt.Errorf("%s %q: rolls %d-%d land on nothing — every roll must be covered "+
				"(to mean \"nothing happens\", say so with an explicit entry)",
				what, id, next, r.Min-1)
		}
		if r.Min < next {
			overlapEnd := r.Max
			if next-1 < overlapEnd {
				overlapEnd = next - 1
			}
			return fmt.Errorf("%s %q: rolls %d-%d are claimed twice (%s overlaps the range before it)",
				what, id, r.Min, overlapEnd, r.Label)
		}
		next = r.Max + 1
	}
	if next <= maxRoll {
		return fmt.Errorf("%s %q: rolls %d-%d land on nothing — every roll must be covered "+
			"(to mean \"nothing happens\", say so with an explicit entry)",
			what, id, next, maxRoll)
	}
	return nil
}

// pickRange returns the index of the range containing roll, or -1. Coverage is
// validated at load, so -1 is unreachable for a loaded def — callers still guard.
func pickRange(ranges []rollRange, roll int) int {
	for i, r := range ranges {
		if roll >= r.Min && roll <= r.Max {
			return i
		}
	}
	return -1
}

File: client/src/game-portal/src/game/items/rollCoverage.ts
// Client mirror of the server's roll-coverage rule (roll_coverage.go).
//
// A weighted list's entries and a table's rows must TILE the die — cover
// 1..maxRoll with no gaps and no overlaps — so every roll lands on exactly one
// outcome. The server re-validates; this exists so an author sees the hole while
// they are making it rather than after clicking Save.
//
// Gaps are an ERROR, not an implicit "nothing happens": a hole in the ranges is
// indistinguishable from a typo, which is exactly how the one gap that ever
// shipped (a deliberate 10% no-drop) read.

/** One claimed slice of the die, labelled by what it yields. */
export type CoverageRange = {
  min: number
  max: number
  label: string
}

/** A contiguous slice of the die, resolved: what it yields and how likely. */
export type CoverageBand = {
  min: number
  max: number
  label: string
  /** Rolls this band covers. */
  rolls: number
  /** Share of the die, 0-100, rounded to one decimal. */
  percent: number
  /** True for a band nothing claims — always an error. */
  uncovered: boolean
}

export type CoverageReport = {
  bands: CoverageBand[]
  errors: string[]
  /** True when the die is completely and uniquely covered. */
  complete: boolean
}

/**
 * Resolves the die into bands and reports what is wrong with it.
 *
 * Returns bands for the WHOLE die, including the uncovered stretches — so the
 * editor can show the hole rather than just say there is one.
 */
export function analyzeCoverage(maxRoll: number, ranges: CoverageRange[]): CoverageReport {
  const errors: string[] = []

  if (!Number.isFinite(maxRoll) || maxRoll < 1) {
    return { bands: [], errors: ['Max roll must be at least 1.'], complete: false }
  }
  if (ranges.length === 0) {
    return { bands: [], errors: ['Needs at least one entry.'], complete: false }
  }

  for (const r of ranges) {
    if (r.min < 1 || r.max > maxRoll) {
      errors.push(`${r.label} claims rolls ${r.min}–${r.max}, outside the die 1–${maxRoll}.`)
    }
    if (r.min > r.max) {
      errors.push(`${r.label} claims rolls ${r.min}–${r.max}, which is backwards.`)
    }
  }

  const sorted = [...ranges].sort((a, b) => a.min - b.min)
  const bands: CoverageBand[] = []
  const band = (min: number, max: number, label: string, uncovered: boolean): CoverageBand => {
    const rolls = max - min + 1
    return { min, max, label, rolls, percent: Math.round((rolls / maxRoll) * 1000) / 10, uncovered }
  }

  let next = 1
  for (const r of sorted) {
    if (r.min > next) {
      bands.push(band(next, r.min - 1, 'nothing — uncovered', true))
      errors.push(`Rolls ${next}–${r.min - 1} land on nothing. Every roll must be covered.`)
    } else if (r.min < next) {
      errors.push(`Rolls ${r.min}–${Math.min(r.max, next - 1)} are claimed twice (${r.label} overlaps the range before it).`)
    }
    if (r.min <= r.max) {
      bands.push(band(Math.max(r.min, 1), Math.min(r.max, maxRoll), r.label, false))
    }
    next = Math.max(next, r.max + 1)
  }
  if (next <= maxRoll) {
    bands.push(band(next, maxRoll, 'nothing — uncovered', true))
    errors.push(`Rolls ${next}–${maxRoll} land on nothing. Every roll must be covered.`)
  }

  return { bands, errors, complete: errors.length === 0 }
}
